pkg/filetree/depth_first_path_walker_test.go

```diff
@@ -0,0 +1,287 @@
+		}
+		possiblePaths[p] = ref
+	}
+
+	// absolute symlink
+	_, err := tr.AddSymLink("/home/elsewhere/symlink", "/sym-linked-dest")
+	if err != nil {
+		t.Fatalf("could not setup link: %+v", err)
+	}
+	// since we are following base links on walk, we should NOT expect the symlink ref at the link destination
+	possiblePaths["/home/elsewhere/symlink"] = nil
+	possiblePaths["/home/elsewhere/symlink/another/a-.gif"] = possiblePaths["/sym-linked-dest/another/a-.gif"]
+
+	// relative symlink
+	_, err = tr.AddSymLink("/home/again/symlink", "../../../sym-linked-dest")
+	if err != nil {
+		t.Fatalf("could not setup link: %+v", err)
+	}
+	// since we are following base links on walk, we should NOT expect the symlink ref at the link destination
+	possiblePaths["/home/again/symlink"] = nil
+	possiblePaths["/home/again/symlink/another/a-.gif"] = possiblePaths["/sym-linked-dest/another/a-.gif"]
+
+	// dead symlink (dir)
+	ref, err := tr.AddSymLink("/home/again/deadsymlink", "../ijustdontexist")
+	if err != nil {
+		t.Fatalf("could not setup link: %+v", err)
+	}
+	possiblePaths["/home/again/deadsymlink"] = ref
+
+	// dead symlink (to txt)
+	ref, err = tr.AddSymLink("/home/again/dead.jpg", "../ialsojustdontexist")
+	if err != nil {
+		t.Fatalf("could not setup link: %+v", err)
+	}
+	possiblePaths["/home/again/dead.jpg"] = ref
+
+	// hardlink
+	ref, err = tr.AddHardLink("/home/elsewhere/hardlink", "/hard-linked-dest")
+	if err != nil {
+		t.Fatalf("could not setup link: %+v", err)
+	}
+	// since we are following base links on walk, we should NOT expect the symlink ref at the link destination
+	possiblePaths["/home/elsewhere/hardlink"] = nil
+	possiblePaths["/home/elsewhere/hardlink/something/b-.gif"] = possiblePaths["/hard-linked-dest/something/b-.gif"]
+
+	// add all paths which should not have references
+	for p := range possiblePaths {
+		for _, c := range file.Path(p).ConstituentPaths() {
+			if _, exists := possiblePaths[string(c)]; !exists {
+				possiblePaths[string(c)] = nil
+			}
+		}
+	}
+
+	return tr, possiblePaths
+}
+
+func TestDFS_WalkAll(t *testing.T) {
+	tr, possiblePaths := dfsTestTree(t)
+
+	actualPaths := make(map[string]*file.Reference, 0)
+	visitor := func(path file.Path, node filenode.FileNode) error {
+		actualPaths[string(path)] = node.Reference
+		return nil
+	}
+
+	walker := NewDepthFirstPathWalker(tr, visitor, nil)
+	if err := walker.WalkAll(); err != nil {
+		t.Fatalf("could not walk: %+v", err)
+	}
+
+	assertExpectedTraversal(t, possiblePaths, actualPaths)
+}
+
+func TestDFS_WalkAll_EarlyTermination(t *testing.T) {
+	tr, possiblePaths := dfsTestTree(t)
+
+	// delete paths we aren't expecting
+	tailPaths := []string{
+		"/place/example.gif",
+		"/sym-linked-dest/another/a-.gif",
+	}
+	for _, p := range tailPaths {
+		for _, c := range file.Path(p).ConstituentPaths() {
+			if c == "/" {
+				continue
+			}
+			delete(possiblePaths, string(c))
+		}
+		delete(possiblePaths, p)
+	}
+	// </delete>
```

Comment:
left over comment? or actionable?

Reply:
ooooh never mind, I think this means that this whole block is for deleting. Perhaps makes sense to extract it out? (nit for sure)

Reply:
this could have been more clear... the above code deletes paths that are not expected from the expected test fixture.... I'll update the comment